nifi-framework-bundle/nifi-framework-extensions/nifi-py4j-framework-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/componentstate.py

```diff
@@ -0,0 +1,171 @@
+        """
+        try:
+            self._java_state_manager.replace(oldValue.get_java_statemap(), convert_pydict_to_jmap(newValue), convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+    def clear(self, scope):
+        """
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        """
+        try:
+            self._java_state_manager.clear(convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+
+class StateMap:
+    """
+    Python class wrapping the StateMap Java implementation.
+
+    The constructor receives a Java object implementing the Java StateMap interface.
+    The methods in Java StateMap have corresponding methods in this class.
+    Methods in this class propagate the method calls to the Java StateMap object.
+    """
+
+    _java_statemap = None
+
+    def __init__(self, java_statemap):
+        """
+        :param java_statemap: Java StateMap object.
+        """
+        self._java_statemap = java_statemap
+
+    def getStateVersion(self):
+        """
+        :return: (int) Integer value of the state version, or None if it is not set.
+        """
+        if self._java_statemap.getStateVersion().isEmpty():
+            return None
+        else:
+            return self._java_statemap.getStateVersion().get()
+
+    def get(self, key):
+        """
+        :param key: (str) String key indicating which state value should be returned.
+        :return: (str) String state value.
+        """
+        return self._java_statemap.get(key)
+
+    def toMap(self):
+        """
+        :return: (dict) Python dictionary representation of the state.
+        """
+        return self.to_dictionary()
+
+    def to_dictionary(self):
+        """
+        :return: (dict) Python dictionary representation of the state.
+        """
+        result = dict()
+
+        java_iterator = self._java_statemap.toMap().entrySet().iterator()
+        while java_iterator.hasNext():
+            java_map_entry = java_iterator.next()
+            result[java_map_entry.getKey()] = java_map_entry.getValue()
+
+        return result
+
+    def get_java_statemap(self):
+        """
+        :return: The Java StateMap object behind this wrapper class.
+        """
+        return self._java_statemap
+
+
+class StateException(Exception):
+    """
+    Exception class for any exception produced by the operations in StateManager.
+
+    The StateManager Python class propagates the method calls to the underlying Java StateManager object.
+    When an operation fails on the Java side, a Py4JJavaError is produced on the Python side by the Py4J bridge.
+    This Py4JJavaError is caught in the Python StateManager and raised as a StateException instead,
+    indicating that the exception carries information regarding an error when accessing the state.
```

Comment:
This comment seems more appropriate on the State Manager class, as opposed to the exception itself.